clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -228,13 +228,16 @@ private void process(final ErrorEvent event) {
         }
 
         private void process(final ConsumerRebalanceListenerCallbackNeededEvent event) {
-            ApplicationEvent invokedEvent = invokeRebalanceCallbacks(
+            ConsumerRebalanceListenerCallbackCompletedEvent invokedEvent = invokeRebalanceCallbacks(
                 rebalanceListenerInvoker,
                 event.methodName(),
                 event.partitions(),
                 event.future()
             );
             applicationEventHandler.add(invokedEvent);
+            if (invokedEvent.error().isPresent()) {
+                throw invokedEvent.error().get();
```

Comment:
I was expecting to see here the logic for wrapping the callback error into a `KafkaException`, but I see it is at a lower level in the `invokeRebalanceCallbacks`, which it's a bit more obfuscated I would say? Still I see how it's deeply tied to the `ConsumerRebalanceListenerCallbackCompletedEvent` so ok for me to leave as it is if we feel it's clear enough.

Reply:
`ConsumerRebalanceListenerCallbackNeededEvent` handles 'assign', 'revoke', and 'lose' callbacks. It was my understanding—I could be wrong—that we wanted to _wait_ to throw the exception after the reconciliation was fully processed. That is, not necessarily on the first callback 🤔

Reply:
However, this implementation has the interesting property that it will both throw the exception _and_ continue processing. It seems like this could potentially yield **two** exceptions, if, say, both the `onPartitionsRevoked()` _and_ `onPartitionsAssigned()` threw exceptions. Is that the intent?

Reply:
Looking at the reconciliation logic, I think if `onPartitionsRevoked` throws, we'll not execute `onPartitionsAssigned`. And the call to `onPartitionsLost` seems to be independent of reconciliation. So not sure how we'd end up with two exceptions.

You are right that there is a behavioral difference around finishing the reconciliation. The old consumer throws _after_ finishing the reconciliation, while the new consumer throws on a different thread, so there is no strict time ordering between finishing the reconciliation and throwing. But I'm struggling to see how one can observe the difference. The reconciliation will have finished the next time the background thread processes any events, so in a sense, you cannot observe the difference based on the queue architecture. The difference may only be observable through shared state that breaks the queue-based architecture. SubscriptionState comes to mind here. Thinking of something like

1. application thread enters poll, fails during rebalance listener execution and throws
2. application thread somehow reads subscription state
3. background thread updates subscription state as part of reconciliation

Now the application thread has observed an "incomplete reconciliation". But after a listener execution has failed, we don't seem to update the subscription state in the reconciliation. 

So in summary - not sure if we are going to notice the different behaviors?

Reply:
@kirktrue  let me know if you are okay with this, otherwise maybe we can discuss it on Thursday

Reply:
Sorry for the delay!

Yes, I'm in agreement with the perspectives you and @lianetm stated. No qualms from me 😄